row.py

```diff
@@ -1048,49 +1048,77 @@ def filter_results(previous_results_file, out_dir):
     mask = ~working_df["text"].str.contains(r"^\d") == True
     working_df.loc[mask, "keep"] = "no"
     working_df.loc[mask, "nondigit_start"] = "fail"
     logging.info("Number parcels starting with a non-digit flagged: %i", mask.value_counts()[1])
 
     #: flag parcels with a ':P' pattern; permits, not relevant - but PUE is okay!
+    #: 123:PUE = pass
+    #: 123:P = fail
+    #: 123:PR = fail
     working_df["permit"] = "pass"
     mask = working_df["text"].str.contains(r"(?!:PUE):P") == True
     working_df.loc[mask, "keep"] = "no"
     working_df.loc[mask, "permit"] = "fail"
     logging.info("Number of permits (:P pattern, but not :PUE) flagged: %i", mask.value_counts()[1])
 
     #: flag parcels with a colon, that do not have a number before the colon
+    #: most of these also fail the nondigit_start filter
+    #: 123:2A = pass
+    #: ABC:2A = fail
     working_df["no_number_before_colon"] = "pass"
     mask = working_df["text"].str.contains(r"^\D*:") == True
     working_df.loc[mask, "keep"] = "no"
     working_df.loc[mask, "no_number_before_colon"] = "fail"
     logging.info("Number of colons not preceeded by a number flagged: %i", mask.value_counts()[1])
 
     #: flag parcels longer than 13 characters
+    #: 1649NT:STPUEQ = pass
+    #: 1649NTS:STPUEQR = fail
     working_df["too_long"] = "pass"
     mask = working_df["text"].str.len() > 13
     working_df.loc[mask, "keep"] = "no"
     working_df.loc[mask, "too_long"] = "fail"
     logging.info("Number of parcels exceeding 13 characters flagged: %i", mask.value_counts()[1])
 
     #: flag parcels with 4 or more letters given that no colon exists (or 1 or 8)
+    #: the colon is often confused as a 1 or 8, so if those exist, the parcel is kept (UDOT prefers this)
+    #: 1572NT:2EC = pass (has a colon)
+    #: 3:ST2EQ = pass (has a colon)
+    #: 31ST2EQ = pass (1 might be a confused colon)
+    #: 3ST2EQ = fail (4 letters and no colon, 1, or 8)
+    #: 4BARS = fail
+    #: 25DBIA = fail
     working_df["too_many_letters"] = "pass"
     mask = (working_df["text"].str.contains("^[^:18]*$")) & (working_df["text"].str.contains("(?:[A-Z][^A-Z]*){4,}"))
     working_df.loc[mask, "keep"] = "no"
     working_df.loc[mask, "too_many_letters"] = "fail"
     logging.info("Number of parcels with 4+ letters and no colon, 1, or 8 flagged: %i", mask.value_counts()[1])
 
-    #: remove parcels with 5 or more numbers in a row (and not a 1 or a 8)
+    #: remove parcels with 5 or more numbers in a row (and not a colon, 1, or a 8)
+    #: the colon is often confused as a 1 or 8, so if those exist, the parcel is kept (UDOT prefers this)
+    #: 1572NT:2EC = pass
+    #: 23582A = pass (8 might be a confused colon)
+    #: 2572:2A = pass (has a colon)
+    #: 257212A = pass (1 might be a confused colon)
+    #: 03702M = fail
+    #: 699062ON = fail
     working_df["five_number_run"] = "pass"
     mask = (working_df["text"].str.contains("^[^:18]*$")) & (working_df["text"].str.contains(r"\d{5,}"))
```

Comment:
**question**: i'm not sure if this  `^[^:18]*$` is working as it's supposed to. i think i need an explanation or some tests. What you're saying is that the 1 and the 8 are sometimes misplaced with the colon and the colon is "always" in the middle of the string. I don't think this expression cares about the placement of the 1 or the 8. Would it make more sense to replace the 1 or the 8 with a `:` and then run it through the other tests?

Reply:
The colon is sometimes OCR'd incorrectly as a 1 or 8.  So UDOT wanted me to treat a 1 or 8 the same as a colon, to be conservative and keep more 'parcels'. You're correct that the expression doesn't care about placement.  I tried to take the simpler/more conservative approach of keeping the string if it contains a colon, 1 or, 8.  We don't have very much time left, so I didn't want to go down too many rabbit holes by adding a bunch of complexity to capture a handful of edge cases.  This filter only flags 223 parcels as it is.

Reply:
Ok I'm good with that. Let's do our best regardless of any arbitrary time estimates.